tests/classes/StringTest.gd

```diff
@@ -0,0 +1,52 @@
+extends MainLoop
+
+var test_string
+
+func _init():
+	test_string = 'Lorem Ipsum'
+
+func _idle(delta):
+	test_withs()
+	test_finds()
+	
+	# OS.exit() no such thing
+	# OS.set_exit_code() doesn't also exit
+	# quit() is SceneTree not MainLoop
+	finish() # this doesn't exit either :(
```

Comment:
Try `return true`?

Reply:
Afraid not.

```
todd@todd-ab28d3:~/repos/remote/brainsick-godot-tests$ git diff
diff --git a/tests/classes/StringTest.gd b/tests/classes/StringTest.gd
index ef7f444..eb8af87 100644
--- a/tests/classes/StringTest.gd
+++ b/tests/classes/StringTest.gd
@@ -2,7 +2,7 @@ extends MainLoop
 
 var test_string
 
-func _init():
+func _initialize():
        test_string = 'Lorem Ipsum'
 
 func _idle(delta):
@@ -13,6 +13,7 @@ func _idle(delta):
        # OS.set_exit_code() doesn't also exit
        # quit() is SceneTree not MainLoop
        finish() # this doesn't exit either :(
+       return true
 
 func _finalize():
        test_string = null
```

Reply:
Well, does it work when using `idle` instead of `_idle`?

Reply:
I would expect that to throw me into an infinite loop.  I call idle() it calls _idle(), fun ensues.  Did you try it?

```
todd@todd-ab28d3:~/Downloads$ cat temp/main.gd
extends MainLoop

func _idle(delta):
	finish()
	idle()
todd@todd-ab28d3:~/Downloads$ ./Godot3.0Beta2/Godot_v3.0-beta2_x11.64 -s temp/main.gd
No touch devices found
OpenGL ES 3.0 Renderer: AMD Radeon (TM) R9 Fury Series (AMD FIJI / DRM 3.18.0 / 4.13.0-21-generic, LLVM 5.0.0)
GLES3: max ubo light: 409
GLES3: max ubo reflections: 455, ubo size: 144
ARVR: Registered interface: Native mobile
^C
```

MainLoop, like many other classes in Godot, exposes multiple faces.  The first face is meant to be driven by external code, say the engine itself or another class.  The second face is for that which is driven, the thing implementing MainLoop.

Engine/Class -> MainLoop -> Implementation

int idle(delta) calls void _idle(delta)
void finish() calls void _finalize()

Reply:
I think I found it..
You should return true in `_iteration` (even if it is marked "void", it isn't).
See [`main.cpp:1705`](https://github.com/godotengine/godot/blob/0294887a0ca1b42f9fddb3858d4ee304b3f4d055/main/main.cpp#L1705) and [`main_loop.cpp:96`](https://github.com/godotengine/godot/blob/0294887a0ca1b42f9fddb3858d4ee304b3f4d055/core/os/main_loop.cpp#L96).

Sorry for the hassle, I couldn't test it earlier.

Reply:
That works!  I actually spent time in main_loop.cpp but didn't put that together.  Thank you!

```
todd@todd-ab28d3:~/Downloads$ cat temp/main.gd
extends MainLoop

func _iteration(delta):
	finish()

	# undocumented; need to know basis only
	return true
todd@todd-ab28d3:~/Downloads$ time ./Godot3.0Beta2/Godot_v3.0-beta2_x11.64 -s temp/main.gd
No touch devices found
OpenGL ES 3.0 Renderer: AMD Radeon (TM) R9 Fury Series (AMD FIJI / DRM 3.18.0 / 4.13.0-21-generic, LLVM 5.0.0)
GLES3: max ubo light: 409
GLES3: max ubo reflections: 455, ubo size: 144
ARVR: Registered interface: Native mobile

real	0m1.125s
user	0m0.089s
sys	0m0.021s
todd@todd-ab28d3:~/Downloads$ echo $?
0
```